lib/std/crypto/bcrypt.zig

```diff
@@ -230,18 +180,203 @@ fn strHashInternal(password: []const u8, rounds_log: u6, salt: [salt_length]u8)
+                }
+                var c2 = bin[i];
+                i += 1;
+                c1 |= (c2 >> 4) & 0x0f;
+                b64[j] = alphabet[c1];
+                j += 1;
+                c1 = (c2 & 0x0f) << 2;
+                if (i >= bin.len) {
+                    b64[j] = alphabet[c1];
+                    j += 1;
+                    break;
+                }
+                c2 = bin[i];
+                i += 1;
+                c1 |= (c2 >> 6) & 3;
+                b64[j] = alphabet[c1];
+                b64[j + 1] = alphabet[c2 & 0x3f];
+                j += 2;
+            }
+            debug.assert(j == b64.len);
+        }
+
+        fn decode(bin: []u8, b64: []const u8) EncodingError!void {
+            var i: usize = 0;
+            var j: usize = 0;
+            while (j < bin.len) {
+                const c1 = @intCast(u8, mem.indexOfScalar(u8, alphabet, b64[i]) orelse
+                    return EncodingError.InvalidEncoding);
+                const c2 = @intCast(u8, mem.indexOfScalar(u8, alphabet, b64[i + 1]) orelse
+                    return EncodingError.InvalidEncoding);
+                bin[j] = (c1 << 2) | ((c2 & 0x30) >> 4);
+                j += 1;
+                if (j >= bin.len) {
+                    break;
+                }
+                const c3 = @intCast(u8, mem.indexOfScalar(u8, alphabet, b64[i + 2]) orelse
+                    return EncodingError.InvalidEncoding);
+                bin[j] = ((c2 & 0x0f) << 4) | ((c3 & 0x3c) >> 2);
+                j += 1;
+                if (j >= bin.len) {
+                    break;
+                }
+                const c4 = @intCast(u8, mem.indexOfScalar(u8, alphabet, b64[i + 3]) orelse
+                    return EncodingError.InvalidEncoding);
+                bin[j] = ((c3 & 0x03) << 6) | c4;
+                j += 1;
+                i += 4;
+            }
+        }
+    };
+
+    fn strHashInternal(
+        password: []const u8,
+        salt: [salt_length]u8,
+        params: Params,
+    ) [hash_length]u8 {
+        var ct = bcrypt(password, salt, params);
+
+        var salt_str: [salt_str_length]u8 = undefined;
+        Codec.encode(salt_str[0..], salt[0..]);
+
+        var ct_str: [ct_str_length]u8 = undefined;
+        Codec.encode(ct_str[0..], ct[0 .. ct.len - 1]);
+
+        var s_buf: [hash_length]u8 = undefined;
+        const s = fmt.bufPrint(
+            s_buf[0..],
+            "{s}b${d}{d}${s}{s}",
+            .{ prefix, params.rounds_log / 10, params.rounds_log % 10, salt_str, ct_str },
+        ) catch unreachable;
+        debug.assert(s.len == s_buf.len);
+        return s_buf;
+    }
+};
 
-    var ct_str: [ct_str_length]u8 = undefined;
-    Codec.encode(ct_str[0..], ct[0 .. ct.len - 1]);
+/// Hash and verify passwords using the PHC format.
+const PhcFormatHasher = struct {
+    const alg_id = "bcrypt";
+    const BinValue = phc_format.BinValue;
+
+    const HashResult = struct {
+        alg_id: []const u8,
+        rl: u6,
```

Comment:
Where does this `rl` parameter come from?

Using the PHC format with `bcrypt` is not common. But:

- Python's `passlib` [uses `r`](https://passlib.readthedocs.io/en/stable/lib/passlib.hash.bcrypt_sha256.html) (and pre-hashes the input)
- This JavaScript implementation [uses `r` too](https://www.npmjs.com/package/phc-bcrypt)

So, maybe we should use the same convention (for the lack of a specification)